src/code.cloudfoundry.org/gorouter/registry/registry.go

```diff
@@ -124,88 +134,104 @@ func (r *RouteRegistry) register(uri route.Uri, endpoint *route.Endpoint) route.
 		// release read lock, insertRouteKey() will acquire a write lock.
 		r.RUnlock()
-		pool = r.insertRouteKey(routekey, uri)
+		pool, routeAdded = r.insertRouteKey(routekey, uri)
 		r.RLock()
 	}
 
 	if endpoint.StaleThreshold > r.dropletStaleThreshold || endpoint.StaleThreshold == 0 {
 		endpoint.StaleThreshold = r.dropletStaleThreshold
 	}
 
-	endpointAdded := pool.Put(endpoint)
+	putResult = pool.Put(endpoint)
 	// Overwrites the load balancing algorithm of a pool by that of a specified endpoint, if that is valid.
 	r.SetTimeOfLastUpdate(t)
 
-	return endpointAdded
+	return putResult, routeAdded
 }
 
-// insertRouteKey acquires a write lock, inserts the route key into the registry and releases the write lock.
-func (r *RouteRegistry) insertRouteKey(routekey route.Uri, uri route.Uri) *route.EndpointPool {
+// insertRouteKey acquires a write lock, inserts the route key into the registry and releases the
+// write lock. If a pool already exists it returns that instead.
+func (r *RouteRegistry) insertRouteKey(routekey route.Uri, uri route.Uri) (pool *route.EndpointPool, poolAdded bool) {
 	r.Lock()
 	defer r.Unlock()
 
 	// double check that the route key is still not found, now with the write lock.
-	pool := r.byURI.Find(routekey)
-	if pool == nil {
-		host, contextPath := splitHostAndContextPath(uri)
-		pool = route.NewPool(&route.PoolOpts{
-			Logger:                 r.logger,
-			RetryAfterFailure:      r.dropletStaleThreshold / 4,
-			Host:                   host,
-			ContextPath:            contextPath,
-			MaxConnsPerBackend:     r.maxConnsPerBackend,
-			LoadBalancingAlgorithm: r.DefaultLoadBalancingAlgorithm,
-		})
-		r.byURI.Insert(routekey, pool)
-		r.logger.Info("route-registered", slog.Any("uri", routekey))
-		// for backward compatibility:
-		r.logger.Debug("uri-added", slog.Any("uri", routekey))
+	pool = r.byURI.Find(routekey)
+	if pool != nil {
+		return pool, false
 	}
-	return pool
+
+	host, contextPath := splitHostAndContextPath(uri)
+	pool = route.NewPool(&route.PoolOpts{
+		Logger:                 r.logger,
+		RetryAfterFailure:      r.dropletStaleThreshold / 4,
+		Host:                   host,
+		ContextPath:            contextPath,
+		MaxConnsPerBackend:     r.maxConnsPerBackend,
+		LoadBalancingAlgorithm: r.DefaultLoadBalancingAlgorithm,
+	})
+	r.byURI.Insert(routekey, pool)
+
+	return pool, true
 }
 
 func (r *RouteRegistry) Unregister(uri route.Uri, endpoint *route.Endpoint) {
 	if !r.endpointInRouterShard(endpoint) {
 		return
 	}
 
-	r.unregister(uri, endpoint)
+	endpointRemoved, routeRemoved := r.unregister(uri, endpoint)
 
 	r.reporter.CaptureUnregistryMessage(endpoint)
 
+	// WARNING: this only works because there are only info log statements after this point. If you
+	// extend this function, this probably needs to be refactored.
+	if !r.logger.Enabled(context.Background(), slog.LevelInfo) {
+		return
+	}
+
+	if endpointRemoved {
+		r.logger.Info("endpoint-unregistered", buildSlogAttrs(uri, endpoint)...)
```

Comment:
Maybe you can move logger statements after all if conditions. In the if-condition you can assign what message should be logged like:
```
if endpointRemoved {
    logMsg = "endpoint-unregistered"
} else{
    logMsg = "endpoint-not-unregistered"
}
if routeRemoved {
    logMsg = "route-unregistered"
    r.reporter.CaptureRouteDeleted()
}
...
if r.logger.Enabled(context.Background(), slog.LevelInfo) {
    r.logger.Info(logMsg, buildSlogAttrs(uri, endpoint)...)
}
```

Reply:
We want to log route and endpoint independent of each other so that wouldn't work. I've moved the metric further up.